src/auto-evo/RunResults.cs

```diff
@@ -368,6 +328,155 @@ public long GetPopulationInPatch(Species species, Patch patch)
+                foreach (var patchEntry in populations)
+                {
+                    if (!speciesInPatches.ContainsKey(patchEntry.Key))
+                        speciesInPatches[patchEntry.Key] = new Dictionary<Species, long>();
+
+                    patchEntry.Value.ToList().ForEach(s =>
+                        speciesInPatches[patchEntry.Key].Add(s.Key, s.Value));
+                }
+            }
+
+            return speciesInPatches;
+        }
+
+        /// <summary>
+        ///   Computes the final population of a species and its splitted forms, by patch.
+        /// </summary>
+        /// <remarks>
+        ///   <para>
+        ///     Species are only returned if their population is above 0.
+        ///   </para>
+        /// </remarks>
+        public Dictionary<Patch, Dictionary<Species, long>> GetPopulationsByPatch(Species species,
+            bool resolveMigrations = false, bool resolveSplits = false, bool includeNewSpecies = true)
+        {
+            var speciesInPatches = new Dictionary<Patch, Dictionary<Species, long>>();
+
+            if (!results.ContainsKey(species))
+            {
+                GD.Print("Species ", species.FormattedName, "not found in results, returning empty dictionary.");
+                return speciesInPatches;
+            }
+
+            var speciesResult = results[species];
+
+            // Get natural variations
+            foreach (var patchPopulationEntry in speciesResult.NewPopulationInPatches)
+            {
+                if (!speciesInPatches.ContainsKey(patchPopulationEntry.Key))
+                    speciesInPatches[patchPopulationEntry.Key] = new Dictionary<Species, long>();
+
+                if (patchPopulationEntry.Value > 0)
+                {
+                    if (includeNewSpecies || speciesResult.NewlyCreated == null)
+                    {
+                        speciesInPatches[patchPopulationEntry.Key].Add(
+                            species, patchPopulationEntry.Value);
+                    }
+                }
+            }
+
+            if (resolveSplits)
+            {
+                // Deal with splits-off
+                if (speciesResult.SplitOff != null && speciesResult.SplitOffPatches != null)
+                {
+                    foreach (var patch in speciesResult.SplitOffPatches)
+                    {
+                        speciesInPatches[patch].Remove(species);
+                        speciesInPatches[patch].Add(speciesResult.SplitOff,
+                            speciesResult.NewPopulationInPatches[patch]);
+                    }
+                }
+            }
+
+            if (resolveMigrations)
+            {
+                // Apply migrations
+                foreach (var migration in speciesResult.SpreadToPatches)
+                {
+                    // Only consider valid migrations
+                    if (migration.From == null || migration.To == null)
```

Comment:
I really dislike this silencing of errors.

Reply:
The main reason I didn't bother with errors was that it was a duplicate, for starters. Now that it became central I agree it really should include errors.

Reply:
Also probably good to throw exceptions as we have a serious algorithm bug if a migration has invalid data. See for example RemoveInvalidMigrations, which I put effort into in order to ensure migrations are always valid in the results (well after that step has ran).

Reply:
Added a bunch of errors throwing, especially one to check that you don't move more pop than you can.